_posts/2023-01-14-trivially-relocatable.md

```diff
@@ -0,0 +1,656 @@
+---
+tags:
+- Subspace
+- C++
+---
+
+## Trivially Relocatable Types in C++/Subspace
+
+* TOC
+{:toc}
+
+I read a [blog post](https://quuxplusone.github.io/blog/2023/01/13/embed-and-initializer-lists/)
+about `#embed` and initializer lists today, and was pleasantly surprised to find a reference to
+[P1144R6 Object relocation in terms of move plus destroy](
+https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1144r6.html) which I was previously
+unaware of. The [Subspace library](https://github.com/chromium/subspace) makes use of
+trivially relocating objects when possible, and it's nice to see work continuing toward bringing
+this idea into the core language.
+
+Relocatable objects is a key property of Rust types and allows for efficient code generation. When
+an object is relocatable, it means that the combination of "move to destination" and
+"destroy source" can be combined into a single `memcpy()`. And LLVM loves to optimize
+`memcpy()`s, so this leads to some really good code generation.
+
+In Rust, _all_ moves are relocations. And any use of an lvalue that is not `Copy` will move
+(relocate) the object.
+
+```rs
+#[derive(Copy, Clone)]
+struct Copy(i32);
+
+struct NoCopy(i32);
+
+fn main() {
+    let c = Copy(0);
+    let n = NoCopy(0);
+
+    let _c2 = c;  // Copies `c` via memcpy().    
+    let _n2 = n;  // Moves `n` to `n2` via memcpy().
+    
+    // `n` is no longer accessible, and will not be dropped, aka its
+    // "destructor" will be skipped in C++ terms.
+    // let _n3 = n;  // Does not compile.
+}
+```
+
+C++ makes the story of moving and relocation more tricky in a few ways:
+
+1. When an object `A` is moved from (ie. another object `B` accesses the object `A` as an rvalue
+reference), the `A` object can enter any possible state as defined by `B`. The only hard requirement
+is that it's valid to run `A`'s destructor, and its `operator=()` if it has one. This makes it very

```

Comment:
A great point that I should include that phrase, as it will help connect people who are familiar with it. I do want to highlight more on the understanding of what code will do rather than just what is legal via the spec too.